File: DataStructure-Practice/Session10/Java/AsteroidCollision.java
https://leetcode.com/problems/asteroid-collision/

class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        
        Stack<Integer> stack = new Stack<>();
        for (int i : asteroids) {
            if (stack.isEmpty() || i > 0 || stack.peek() < 0) {
                stack.push(i);
            } else {
                boolean destroyed = false;
                while (!stack.isEmpty() && stack.peek() > 0 && !destroyed) {
                    int top = stack.peek();
                    if (Math.abs(i) > top) {
                        stack.pop();
                    } else if (Math.abs(i) < top) {
                        destroyed = true;
                    } else {
                        stack.pop();
                        destroyed = true;
                    }
                }
                if (!destroyed) {
                    stack.push(i);
                }
            }
        }
        
        int[] result = new int[stack.size()];
        int index = stack.size() - 1;
        while (!stack.isEmpty()) {
            result[index--] = stack.pop();
        }

        return result;
        
    }
}

File: DataStructure-Practice/Session10/C++/AsteroidCollision.cpp
https://leetcode.com/problems/asteroid-collision/

class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        // collision => s.top() (+) && i (-) 
        stack<int> s;
        for(auto i:asteroids) {
            if(s.empty() || i>0 || s.top()<0) {
                s.push(i);
            }
            else {
                bool destroyed = false;
                while(!s.empty() && s.top()>0 && !destroyed) {
                    int tp = s.top();
                    if(abs(i) > tp) {
                       s.pop(); 
                    }
                    else if(abs(i) < tp) {
                        destroyed = true;
                    }
                    else {
                        s.pop();
                        destroyed = true;
                    }
                }
                if(!destroyed) {
                    s.push(i);
                }
            }
        }
        
        vector<int> ans(s.size());
        int index = s.size()-1;
        while (!s.empty()) {
            ans[index--] = s.top();
            s.pop();
        }

        return ans;
        
    }
};